### PR TITLE
[LWG 29] P2630R4 submdspan

### LINKED ISSUE #6315: [2023-06 LWG Motion 29] P2630R4 submdspan
https://wiki.edg.com/pub/Wg21varna/StrawPolls/P2630R4.html

### PR DESCRIPTION
**Pull request by burblebee**
Fixes #6315.
Fixes cplusplus/papers#1293

### CODE REVIEW

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/containers.tex [hunk 1]
```diff
@@ -19598,6 +19641,17 @@
 
   private:
     extents_type @\exposid{extents_}@{};    // \expos
+
+    // \ref{mdspan.submdspan.mapping}, \tcode{submdspan} mapping specialization
+    template<class... SliceSpecifiers>
+      constexpr auto @\exposid{submdspan-mapping-impl}@(                    // \expos
+        SliceSpecifiers ... slices) const -> @\seebelow@;
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:19648]: thanks for all these!

File: source/containers.tex [hunk 2]
```diff
@@ -20957,3 +21022,628 @@
...
+// zero out just the surface
+template<class T, class E, class L, class A>
+void zero_surface(mdspan<T, E, L, A> grid3d) {
+  static_assert(grid3d.rank() == 3);
+  zero_2d(submdspan(grid3d, 0, full_extent, full_extent));
+  zero_2d(submdspan(grid3d, full_extent, 0, full_extent));
+  zero_2d(submdspan(grid3d, full_extent, full_extent, 0));
+  zero_2d(submdspan(grid3d, grid3d.extent(0)-1, full_extent, full_extent));
+  zero_2d(submdspan(grid3d, full_extent, grid3d.extent(1)-1, full_extent));
+  zero_2d(submdspan(grid3d, full_extent, full_extent, grid3d.extent(2)-1));
```
[the remark this answers is not in the document]
Comment (JohelEGP) [reply source/containers.tex:21645]: Opened #6353 for this.

File: source/containers.tex [hunk 3]
```diff
@@ -20957,3 +21022,628 @@
...
+
+\begin{itemdescr}
+\pnum
+\mandates
+\tcode{IndexType} is a signed or unsigned integer type.
+
+\pnum
+\returns
+An \tcode{array<IndexType, sizeof...(SliceSpecifiers)> src_idx} such that
+for each $k$ in the range \range{0}{sizeof...(SliceSpecifiers)},
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21234]: It was not an accident. The /k/ here is just for the ranging index and has no scope beyond the scope of this returns clause. We tend to use math fonts for indexes like this so I did.

What we have above was left as is because I was confused about which k we were talking about - was it supposed to be a localized index or the parameter?

File: source/containers.tex [hunk 4]
```diff
@@ -20957,3 +21022,628 @@
...
+\item
+\tcode{get<1>($s_k$)}
+if $S_k$ models \tcode{\exposconcept{index-pair-like}<index_type>}; otherwise
+\item
+\tcode{\exposid{de-ice}($s_k$.offset)} \tcode{+}
+\tcode{\exposid{de-ice}($s_k$.extent)}
+if $S_k$ is a specialization of \tcode{strided_slice}; otherwise
+\item
+%FIXME: Note that the paper uses \tcode{k} here.
+\tcode{src.extent(k)}.
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21207]: Yes, because In the _`src-indices`_ case, there is no parameter `k` to confuse it with; it's clear that we mean the ranging index /_k_/ introduced in the /_Returns_/ clause p11 (or the 1st bullet p11.1, but that's a different problem).

File: source/containers.tex [hunk 5]
```diff
@@ -20957,3 +21022,628 @@
...
+
+\pnum
+\returns
+An \tcode{array<IndexType, sizeof...(SliceSpecifiers)> src_idx} such that
+for each $k$ in the range \range{0}{sizeof...(SliceSpecifiers)},
+\tcode{src_idx[$k$]} equals
+
+\begin{itemize}
+\item
+%FIXME: We're redefining what $k$ is here.
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21239]: Great, then I'd like to pick a different index. Should p11.1 then be:
`first_<IndexType, _j_>(slices...) for each _j_ where map-rank [_j_] equals dynamic_extent`?

If so, then we're saying `__map-rank_[_k_]_ `is equal to that, which makes no sense (the 2nd part of my FIXME comment).

Just what is the paper trying to say here? If someone can tell me, I can likely fix the wording editorially. As it is, the wording is nonsense.

File: source/containers.tex [hunk 6]
```diff
@@ -20957,3 +21022,628 @@
...
+  \item
+  if $S_k$ is a specialization of \tcode{strided_slice}
+    \begin{itemize}
+    \item \tcode{$s_k$.extent = 0}, or
+    \item \tcode{$s_k$.stride > 0}
+    \end{itemize}
+  \item
+  \tcode{0 $\le$ \exposid{first_}<index_type, $k$>(slices...)}
+  \tcode{$\le$ \exposid{last_}<$k$>(src.extents(), slices...)}
+  \tcode{$\le$ \linebreak{}src.extent($k$)}
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21586]: Using math font seems fine, will change.
But, looking elsewhere for examples of lists starting with "`all of the following are \tcode{true}`", we don't use full sentences for the items, so it's more consistent as is. That said, the "`, or`" is a bit of a thorn. 

Ok to resolve? reopen if not satisfied.

File: source/containers.tex (hunk 2 above)
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21645]: Thank you for opening an issue for this, resolving for now.

File: source/containers.tex [hunk 7]
```diff
@@ -20957,3 +21022,628 @@
...
+\begin{note}
+These conditions ensure that the mapping returned by \tcode{submdspan_mapping}
+matches the algorithmically expected index-mapping given the slice specifiers.
+\end{note}
+
+\pnum
+\effects
+Equivalent to:
+\begin{codeblock}
+auto sub_map_offset = submdspan_mapping(src.mapping(), args...);
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21612]: done in separate commit

File: source/containers.tex [hunk 8]
```diff
@@ -20957,3 +21022,628 @@
...
+
+\pnum
+\mandates
+For each rank index $k$ of \tcode{src.extents()},
+exactly one of the following is true:
+\begin{itemize}
+\item $S_k$ models \tcode{\libconcept{convertible_to}<IndexType>}
+\item $S_k$ models \tcode{\exposconcept{index-pair-like}<IndexType>}
+\item \tcode{is_convertible_v<$S_k$, full_extent_t>} is \tcode{true}
+\item $S_k$ is a specialization of \tcode{strided_slice}
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21273]: Went with ", ... or" which is more commonly used elsewhere in the spec.

File: source/containers.tex [hunk 9]
```diff
@@ -20957,3 +21022,628 @@
...
+\begin{itemdecl}
+template<class IndexType, class ... Extents, class ... SliceSpecifiers>
+  constexpr auto submdspan_extents(const extents<IndexType, Extents...>& src,
+                                   SliceSpecifiers ... slices);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\constraints
+\tcode{sizeof...(slices)} equals \tcode{Extents::rank()},
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21263]: Punctuation fixed; will revisit use of "rank"...

File: source/containers.tex [hunk 10]
```diff
@@ -20957,3 +21022,628 @@
...
+\begin{itemdescr}
+\pnum
+\tcode{strided_slice} has the data members and special members specified above.
+It has no base classes or members other than those specified.
+
+\pnum
+\mandates
+\tcode{OffsetType}, \tcode{ExtentType}, and \tcode{StrideType}
+are signed or unsigned integer types, or
+model \exposconcept{integral-constant-like}.
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21099]: Leaving as is - this wording is typical of wording used elsewhere in the spec. FYI I could find no cases which use "models of".

File: source/containers.tex [hunk 11]
```diff
@@ -20957,3 +21022,628 @@
...
+\pnum
+\expects
+For each rank index $k$ of \tcode{src.extents()},
+all of the following are \tcode{true}:
+\begin{itemize}
+\item
+if $S_k$ is a specialization of \tcode{strided_slice}
+  \begin{itemize}
+  \item \tcode{$s_k$.extent = 0}, or
+  \item \tcode{$s_k$.stride > 0}
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21285]: Changed to use math font.

I don't know what punctuation to use for this, but having an embedded sentence seems wrong and inconsistent.
We have a list of expressions, each of which might be true... We typically don't use any punctuation, or separate with ", and/or", but the nested or'ed list throws things off.

Got another suggestion?

File: source/containers.tex [hunk 12]
```diff
@@ -20957,3 +21022,628 @@
...
+\item $S_k$ is a specialization of \tcode{strided_slice}
+\end{itemize}
+
+\pnum
+\expects
+For each rank index $k$ of \tcode{src.extents()},
+all of the following are \tcode{true}:
+\begin{itemize}
+\item
+if $S_k$ is a specialization of \tcode{strided_slice}
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21282]: See reply above (about punctuation for this clause)

File: source/containers.tex [hunk 13]
```diff
@@ -20957,3 +21022,628 @@
...
+  \begin{itemize}
+  \item \tcode{$s_k$.extent = 0}, or
+  \item \tcode{$s_k$.stride > 0}
+  \end{itemize}
+\item
+\tcode{0 $\le$ \exposid{first_}<IndexType, $k$>(slices...)}
+\tcode{$\le$ \exposid{last_}<$k$>(src, slices...)}
+\tcode{$\le$ src.extent($k$)}
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21290]: Not sure what is preferred, but changed anyway. End result looks identical, so maybe it doesn't matter?

File: source/containers.tex [hunk 14]
```diff
@@ -20957,3 +21022,628 @@
...
+Let \tcode{SubExtents} be a specialization of \tcode{extents} such that:
+
+\begin{itemize}
+\item
+%FIXME: I think we want the count here, "number" is ambiguous.
+\tcode{SubExtents::rank()} equals the number of $k$ such that
+$S_k$ does not model \tcode{\libconcept{convertible_to}<IndexType>}; and
+
+\item
+for all rank index $k$ of \tcode{Extents} such that
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21303]: Yup. But since there's no parameter `k`, it's clear that we're only talking about the index _k_, so all good here.

File: source/containers.tex [hunk 15]
```diff
@@ -20957,3 +21022,628 @@
...
+  model \exposconcept{integral-constant-like}; otherwise
+
+  \item
+  \tcode{0},
+  if $S_k$ is a specialization of \tcode{strided_slice}, whose
+%FIXME: The following line build fails with:
+%FIXME: ::error concept integral-cons\-tant-like has no definition
+% \tcode{extent_type} models \exposconcept{integral-cons\-tant-like},
+%FIXME: Using \exposid as a workaround:
+  \tcode{extent_type} models \exposid{integral-cons\-tant-like},
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21328]: Thank you!!!

File: source/containers.tex [hunk 16]
```diff
@@ -20957,3 +21022,628 @@
...
+
+  \item
+  \tcode{dynamic_extent}.
+  \end{itemize}
+\end{itemize}
+
+\pnum
+\returns
+A value \tcode{ext} of type \tcode{SubExtents} such that
+for each $k$
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21350]: We want the index _k_ so all good here; correct as is.

File: source/containers.tex [hunk 17]
```diff
@@ -20957,3 +21022,628 @@
...
+if $S_k$ is a specialization of \tcode{strided_slice}, otherwise
+
+\item
+\tcode{ext.extent(\placeholder{map-rank}[$k$])} equals
+\tcode{\exposid{last_}<$k$>(src, slices...) -}
+\tcode{\exposid{first_}<IndexType, $k$>(\linebreak{}slices...)}.
+\end{itemize}
+\end{itemdescr}
+
+\rSec4[mdspan.submdspan.mapping]{Layout specializations of \tcode{submdspan_mapping}}
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21367]: Would need to consult the authors for how to sync these up. Are any of them watching this?

[the remark this answers is not in the document]
Comment (JohelEGP) [reply source/containers.tex:21367]: Doubt it.

File: source/containers.tex [hunk 18]
```diff
@@ -20957,3 +21022,628 @@
...
+
+\pnum
+\mandates
+For each rank index $k$ of \tcode{extents()},
+exactly one of the following is true:
+\begin{itemize}
+\item $S_k$ models \tcode{\libconcept{convertible_to}<index_type>}
+\item $S_k$ models \tcode{\exposconcept{index-pair-like}<index_type>}
+\item \tcode{is_convertible_v<$S_k$, full_extent_t>} is \tcode{true}
+\item $S_k$ is a specialization of \tcode{strided_slice}
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21404]: Used ", or" for consistency.

[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21404]: Keeping as is. Please open an edit issue/pr if you'd like a change.

File: source/containers.tex [hunk 19]
```diff
@@ -20957,3 +21022,628 @@
...
+\item
+if $S_k$ is a specialization of \tcode{strided_slice}
+  \begin{itemize}
+  \item \tcode{$s_k$.extent = 0}, or
+  \item \tcode{$s_k$.stride > 0}
+  \end{itemize}
+\item
+\tcode{0 $\le$ \exposid{first_}<index_type, $k$>(slices...)}
+\tcode{$\le$ \exposid{last_}<$k$>(extents(), slices...)}
+\tcode{$\le$ extents(\linebreak{}).extent($k$)}
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21421]: Changed to use math code, but left punctuation as is - see related situation above.

File: source/containers.tex [hunk 20]
```diff
@@ -20957,3 +21022,628 @@
...
+
+\pnum
+Let \tcode{sub_ext} be the result of
+\tcode{submdspan_extents(extents(), slices...)} and
+let \tcode{SubExtents} be \tcode{decltype(sub_ext)}.
+
+\pnum
+Let \tcode{sub_strides} be
+an \tcode{array<SubExtents::index_type, SubExtents::rank()>} such that
+for each rank index $k$ of \tcode{extents()}
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21432]: _k_ looks correct here

File: source/containers.tex [hunk 21]
```diff
@@ -20957,3 +21022,628 @@
...
+\item
+\tcode{submdspan_mapping_result\{layout_stride::mapping(sub_ext, sub_strides), offset\}}.
+\end{itemize}
+\end{itemdescr}
+
+\rSec4[mdspan.submdspan.submdspan]{\tcode{submdspan} function}
+
+\indexlibraryglobal{submdspan}%
+\begin{itemdecl}
+// \ref{mdspan.submdspan}, \tcode{submdspan} creation
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21515]: Right, this is the itemdecl/descr for `submdspan` which was first declared in is in [mdspan.syn].  Probably a reference to this section was meant to be added there instead. That said, I'm undoing your suggestion to change the section name in the comment but won't unresolve it cuz it no longer applies.

Moving comment to synopsis and fixing reference...

File: source/containers.tex [hunk 22]
```diff
@@ -20957,3 +21022,628 @@
...
+is \tcode{true}, and
+
+\item
+For each rank index $k$ of \tcode{src.extents()},
+exactly one of the following is true:
+  \begin{itemize}
+  \item $S_k$ models \tcode{\libconcept{convertible_to}<index_type>}
+  \item $S_k$ models \tcode{\exposconcept{index-pair-like}<index_type>}
+  \item \tcode{is_convertible_v<$S_k$, full_extent_t>} is \tcode{true}
+  \item $S_k$ is a specialization of \tcode{strided_slice}
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21566]: Used ", or"

**Comment by burblebee** [comment]
The build is failing with:
`::error concept hyperpage1059 has no definition
`
I expect this is due to the two uses of \exposconceptx:
*  `\tcode{extent_type} models \exposconceptx{integral-cons\-tant-like},`
 * `model \\exposconceptx{integral-constant-\linebreak{}like}; otherwise`

How to fix?

File: source/containers.tex [hunk 23]
```diff
@@ -20957,3 +21022,628 @@
...
+\pnum
+\tcode{strided_slice} represents a set of
+\tcode{extent} regularly spaced integer indices.
+The indices start at \tcode{offset}, and
+increase by increments of \tcode{stride}.
+
+\indexlibraryglobal{strided_slice}%
+\begin{itemdecl}
+template<class OffsetType, class ExtentType, class StrideType>
+struct strided_slice {
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21080]: Agreed, fixed

File: source/containers.tex [hunk 24]
```diff
@@ -20957,3 +21022,628 @@
...
+\item let \tcode{index_type} be
+  \begin{itemize}
+  \item
+  \tcode{M::index_type}
+  if the function is a member of a class \tcode{M}, otherwise
+  \item
+  \tcode{remove_reference_t<decltype(src)>::index_type}
+  if the function has a parameter named \tcode{src}, otherwise
+  \item
+  the same type as the function's template argument \tcode{IndexType};
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21051]: Adding in a separate commit

**Comment by burblebee** [comment]
The builds are failing during the system install - who maintains the build machines?

File: source/containers.tex [hunk 25]
```diff
@@ -20957,3 +21022,628 @@
...
+\pnum
+\returns
+\tcode{Extents::\exposid{index-cast}($\lambda{}_k$)}.
+\end{itemdescr}
+
+\indexlibraryglobal{\exposid{src-indices}}%
+\begin{itemdecl}
+template<class IndexType, size_t N, class ... SliceSpecifiers>
+  constexpr array<IndexType, sizeof...(SliceSpecifiers)>
+    @\exposid{src-indices}@(const array<IndexType, N>& indices, SliceSpecifiers ... slices);
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex]: Please open a separate issue for the issues with "rank"; the issues with paper are already overwhelming.

Then, resolve your related comments here so we can reduce the noise level some?

File: source/containers.tex [hunk 26]
```diff
@@ -20957,3 +21023,631 @@
...
+\item
+\tcode{sub_map_offset.mapping.extents() == submdspan_extents(src.mapping(), slices...)}\linebreak
+is \tcode{true}; and
+
+\item
+for each integer pack \tcode{I} which is a multidimensional index
+in \tcode{sub_map_offset.mapping.extents()},
+\begin{codeblock}
+sub_map_offset.mapping(I...) + sub_map_offset.offset ==
+  src.mapping()(@\exposid{src-indices}@(array{I...}, slices ...))
```
Comment (JohelEGP) [inline source/containers.tex:21602]: Missing part of <https://github.com/cplusplus/draft/pull/6338#discussion_r1240037371>.
```suggestion
  src.mapping()(@\exposid{src-indices}@(array{I...}, slices...))
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]
There are some "otherwise" left to be moved to the beginning of the next list item (as per #5810).
I didn't add suggestion for those, before or now, because they're a bit more involved.

Still open comments:
- <https://github.com/cplusplus/draft/pull/6338#discussion_r1239849668>
- <https://github.com/cplusplus/draft/pull/6338#discussion_r1239968386>

Opened #6365 with regards to `rank`.

File: source/containers.tex (hunk 11 above)
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21294]: Oops! Fixed again.

File: source/containers.tex [hunk 27]
```diff
@@ -20957,3 +21023,631 @@
...
+  \tcode{\exposid{de-ice}(tuple_element_t<0, $S_k$>())}
+  if $S_k$ models \tcode{\exposconcept{index-pair-like}<IndexType>}, and
+  both \tcode{tuple_element_t<0, $S_k$>} and
+  \tcode{tuple_elem\-ent_t<1, $S_k$>}
+  model \exposconcept{integral-constant-like}; otherwise
+
+  \item
+  \tcode{0},
+  if $S_k$ is a specialization of \tcode{strided_slice}, whose
+  \tcode{extent_type} models \exposconceptx{integral-cons\-tant-like},
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21333]: Ah, I see. Fixed.

File: source/containers.tex [hunk 28]
```diff
@@ -20957,3 +21023,631 @@
...
+  for which \tcode{extent_type()} equals zero; otherwise
+
+  \item
+  \tcode{1 + (\exposid{de-ice}($S_k$::extent_type()) - 1) /}
+  \tcode{\exposid{de-ice}($S_k$::stride_type())},
+  if $S_k$ is a specialization of \tcode{strided_slice} whose
+  \tcode{extent_type} and \tcode{stride_type}
+  model \exposconceptx{integral-cons\-tant-like};
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex:21341]: Fixed.

File: source/containers.tex (hunk 26 above)
Comment (burblebee) [reply source/containers.tex:21602]: Many thanks, fixed.

**Review by JohelEGP — APPROVED** [review]

File: source/containers.tex [hunk 29]
```diff
@@ -21425,3 +21491,631 @@
...
+namespace std {
+  template<class OffsetType, class ExtentType, class StrideType>
+  struct strided_slice {
+    using offset_type = OffsetType;
+    using extent_type = ExtentType;
+    using stride_type = StrideType;
+
+    [[no_unique_address]] OffsetType offset{};
+    [[no_unique_address]] ExtentType extent{};
+    [[no_unique_address]] StrideType stride{};
```
Comment (tkoeppe) [inline source/containers.tex:21559]: Why are we using the template arguments here instead of the member typedef names?

Comment (jensmaurer) [reply source/containers.tex:21559]: No idea. I think those can/should be fixed editorially.

File: source/containers.tex [hunk 30]
```diff
@@ -21425,3 +21490,631 @@
...
+  };
+}
+\end{codeblock}
+
+\pnum
+\tcode{strided_slice} has the data members and special members specified above.
+It has no base classes or members other than those specified.
+
+\pnum
+\mandates
```
Comment (tkoeppe) [inline source/containers.tex:21568]: Hm, now that we got rid of the itemdecl/itemdescr, it's odd to have a `\mandates` here.

Comment (tkoeppe) [reply source/containers.tex:21568]: @jensmaurer Can we say "OffsetType, ExtentType, and StrideType shall be signed or unsigned integer types, or model integral-constant-like; otherwise the program is ill-formed."?

File: source/containers.tex [hunk 31]
```diff
@@ -21425,3 +21490,631 @@
...
+\tcode{sizeof...(slices)} equals \tcode{Extents::rank()}.
+
+\pnum
+\mandates
+For each rank index $k$ of \tcode{src.extents()},
+exactly one of the following is true:
+\begin{itemize}
+\item $S_k$ models \tcode{\libconcept{convertible_to}<IndexType>},
+\item $S_k$ models \tcode{\exposconcept{index-pair-like}<IndexType>},
+\item \tcode{is_convertible_v<$S_k$, full_extent_t>} is \tcode{true}, or
```
Comment (tkoeppe) [inline source/containers.tex:21748]: Where is `full_extent_t` defined?

File: source/containers.tex (hunk 25 above)
[the remark this answers is not in the document]
Comment (tkoeppe) [reply source/containers.tex]: I agree, I'd appreciate a separate issue/PR. It would definitely be nice to simplify the wording, but let's keep that out of the immediate paper review.

File: source/containers.tex (hunk 31 above)
Comment (JohelEGP) [reply source/containers.tex:21748]: I opened #6353 for this.
P0009 (R18) mentions in "1.8 P0009r11: 2021-05 Mailing"
> Renamed all_type to full_extent_t and all to full_extent

My guess is that `submdspan` wasn't up to date.

File: source/containers.tex (hunk 25 above)
[the remark this answers is not in the document]
Comment (JohelEGP) [reply source/containers.tex]: Yes, that's #6365.

File: source/containers.tex (hunk 7 above)
[the remark this answers is not in the document]
Comment (tkoeppe) [reply source/containers.tex]: You could send a PR for this; sounds like a nice simplification.

File: source/containers.tex (hunk 29 above)
Comment (tkoeppe) [reply source/containers.tex]: Done.

File: source/containers.tex (hunk 31 above)
Comment (tkoeppe) [reply source/containers.tex:21748]: @crtrott ^^^ ?

Comment (tkoeppe) [reply source/containers.tex:21748]: OK, thanks for the issue! We can defer that if there's no immediate solution at hand.

File: source/containers.tex (hunk 30 above)
Comment (tkoeppe) [reply source/containers.tex:21568]: Filed #6397.

File: source/containers.tex (hunk 7 above)
[the remark this answers is not in the document]
Comment (JohelEGP) [reply source/containers.tex]: Opened #6399 for this.
